### PR TITLE
prevent bi-directional connections from triggering duplicate-ack fast-retransmits

### PR DESCRIPTION
**Pull request by arvidn**

### CODE REVIEW

**Comment by arvidn** [comment]
I discovered this problem in the libtorrent implementation and figured I'd check libutp as well. Here's a reference on what BSD does (or at least did): https://www.ietf.org/mail-archive/web/tcpm/current/msg01200.html

File: utp_internal.cpp [hunk 1]
```diff
@@ -1903,10 +1903,21 @@ size_t utp_process_incoming(UTPSocket *conn, const byte *packet, size_t len, boo
 
 	// if we get the same ack_nr as in the last packet
 	// increase the duplicate_ack counter, otherwise reset
-	// it to 0
+	// it to 0.
+	// It's important to only count ACKs in ST_STATE packets. Any other
```
Comment (jpcottin) [inline utp_internal.cpp]: agreed

File: utp_internal.cpp [hunk 2]
```diff
@@ -1903,10 +1903,21 @@ size_t utp_process_incoming(UTPSocket *conn, const byte *packet, size_t len, boo
...
+	// a bittorrent HAVE message), we're very likely to see 3 duplicate ACKs
+	// immediately after sending our payload packet. This effectively disables
+	// the fast-resend on duplicate-ack logic for bi-directional connections
+	// (except in the case of a selective ACK). This is in line with BSD4.4 TCP
+	// implementation.
 	if (conn->cur_window_packets > 0) {
 		if (pk_ack_nr == ((conn->seq_nr - conn->cur_window_packets - 1) & ACK_NR_MASK)
-			&& conn->cur_window_packets > 0) {
+			&& conn->cur_window_packets > 0
+			&& pk_flags == ST_STATE) {
```
Comment (jpcottin) [inline utp_internal.cpp]: ok, lgtm

**Comment by B-C** [comment]
lgtm

**Comment by arvidn** [comment]
some more background, the case where the current logic breaks (as in causes an unnecessary packet loss and resend) is when we're receiving at a reasonably high rate (with more than 3 packets in-flight from the other end) and we decide to send a single payload packet. Immediately following, every packet we receive will count as a duplicate ack (unless this patch is applied) and the packet we sent out will be considered dropped within a full round-trip

**Pull request merged**